File: haggis-platform/backend/src/Haggis.AI/HeuristicEngine/HeuristicOptions.cs
namespace Haggis.AI.Strategies
{
    public sealed class HeuristicOptions
    {
        // All options below are external multipliers for strategy base scores.
        // 0 disables a strategy, 1 keeps its built-in default strength, values
        // above 1 strengthen its effect proportionally.

        // PreferSinglesNotBreakingNonWildCombinationsWeightStrategy
        public float PreferSinglesNotBreakingNonWildCombinationsWeight { get; set; } = 1f;

        // PreferTricksWithMoreContinuationsWeightStrategy
        public float ContinuationCountWeight { get; set; } = 2.455f;

        // PreferTricksThatAreMostLikelyNonBreakableWeightStrategy
        public float PreferNonBreakableOpeningWeight { get; set; } = 0f;

        // PreferLowerTricksWhenHandIsLargeWeightStrategy
        public float PreferLowerStartWeight { get; set; } = 1.983f;

        // PreferShorterTricksWhenHandIsLargeWeightStrategy
        public float PreferShorterStartWeight { get; set; } = 1.686f;

        // PenalizeWildCardsInOpeningWeightStrategy
        public float WildCardOpeningWeight { get; set; } = 4.418f;

        // PenalizeBombOpeningWeightStrategy
        public float BombOpeningWeight { get; set; } = 0.608f;

        // PenalizeOpeningWhenHigherRelatedCombinationExistsWeightStrategy
        public float HigherRelatedCombinationOpeningWeight { get; set; } = 3.458f;

        // PenalizeWildCardsInContinuationWeightStrategy
        public float WildCardContinuationWeight { get; set; } = 0f;

        // PenalizeBombContinuationWeightStrategy
        public float BombContinuationWeight { get; set; } = 3.646f;

        // PenalizeContinuationWhenHigherRelatedCombinationExistsWeightStrategy
        public float HigherRelatedCombinationContinuationWeight { get; set; } = 0.224f;

        // PreferUsingWildAsHigherCardInContinuationWeightStrategy
        public float PreferUsingWildAsHigherCardInContinuationWeight { get; set; } = 1.528f;

        // PreferLowerValueContinuationWeightStrategy
        public float LowerValueContinuationWeight { get; set; } = 3.652f;

        // PreferContinuationsWithFollowUpWeightStrategy
        public float ContinuationFollowUpWeight { get; set; } = 1f;

        // PreferNotPassingWhenHoldingPlayableBombInEndgameWeightStrategy
        public float PlayableBombInEndgameWeight { get; set; } = 0f;
    }
}

File: haggis-platform/client/Haggis.Flutter.UI/lib/local_game/ai/heuristic_options.dart
/// Exact Dart counterpart of Haggis.AI.Strategies.HeuristicOptions.
class HeuristicOptions {
  const HeuristicOptions({
    this.preferSinglesNotBreakingNonWildCombinationsWeight = 1,
    this.continuationCountWeight = 2.455,
    this.preferNonBreakableOpeningWeight = 0,
    this.preferLowerStartWeight = 1.983,
    this.preferShorterStartWeight = 1.686,
    this.wildCardOpeningWeight = 4.418,
    this.bombOpeningWeight = 0.608,
    this.higherRelatedCombinationOpeningWeight = 3.458,
    this.wildCardContinuationWeight = 0,
    this.bombContinuationWeight = 3.646,
    this.higherRelatedCombinationContinuationWeight = 0.224,
    this.preferUsingWildAsHigherCardInContinuationWeight = 1.528,
    this.lowerValueContinuationWeight = 3.652,
    this.continuationFollowUpWeight = 1,
    this.playableBombInEndgameWeight = 0,
  });

  final double preferSinglesNotBreakingNonWildCombinationsWeight;
  final double continuationCountWeight;
  final double preferNonBreakableOpeningWeight;
  final double preferLowerStartWeight;
  final double preferShorterStartWeight;
  final double wildCardOpeningWeight;
  final double bombOpeningWeight;
  final double higherRelatedCombinationOpeningWeight;
  final double wildCardContinuationWeight;
  final double bombContinuationWeight;
  final double higherRelatedCombinationContinuationWeight;
  final double preferUsingWildAsHigherCardInContinuationWeight;
  final double lowerValueContinuationWeight;
  final double continuationFollowUpWeight;
  final double playableBombInEndgameWeight;
}
